toolkit/offchain/src/collateral_selection.rs

```diff
@@ -0,0 +1,77 @@
+use ogmios_client::types::*;
+
+pub const MIN_REQUIRED_COLLATERAL: u64 = 5_000_000;
+
+// Largest-First selection algorithm based on https://cips.cardano.org/cip/CIP-0002#largest-first
+pub fn largest_first(
+	inputs_available: &[OgmiosUtxo],
+	amount_required: u64,
+	max_collateral_inputs: u32,
+	min_required_collateral: u64,
+) -> Result<Vec<&OgmiosUtxo>, String> {
+	let amount_required = amount_required.max(min_required_collateral);
+	let amount_available: u64 = inputs_available.iter().map(|utxo| utxo.value.lovelace).sum();
+	if amount_available < amount_required {
+		return Err(format!("The available amount ({amount_available}) is less than the required amount ({amount_required})"));
+	}
+	let mut inputs_available_sorted = inputs_available.iter().collect::<Vec<_>>();
+	inputs_available_sorted
+		.sort_by(|utxo_a, utxo_b| utxo_a.value.lovelace.cmp(&utxo_b.value.lovelace));
+
+	let mut inputs_selected = Vec::new();
+	let mut sum = 0;
+	while sum < amount_required {
+		if let Some(utxo) = inputs_available_sorted.pop() {
+			sum += utxo.value.lovelace;
+			inputs_selected.push(utxo);
+		}
+	}
+
+	if inputs_selected.len() as u32 > max_collateral_inputs {
+		return Err(format!("Could not find required amount ({amount_required}) of lovelace in the maximum allowed inputs ({max_collateral_inputs})"));
```

Comment:
Please consider:
"Could not find required amount 5000000 of lovelace in the maximum allowed inputs 3" 
vs
"Could not find 50000000 lovelace required for collateral in the 3 maximum allowed inputs"